### PR TITLE
Fix unsafe publication in opt-out query cache

### PR DESCRIPTION
**Pull request by jasontedor**
This opt-out query cache has an unsafe publication issue, where the cache is exposed to another thread (namely the cluster state update thread) before the constructor has finished execution. This exposes the opt-out query cache to concurrency bugs. This commit addresses this by ensuring that the opt-out query cache is not registered as a listener for license state changes until after the constructor has returned.

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java [hunk 1]
```diff
@@ -44,6 +44,14 @@ public OptOutQueryCache(
...
         this.licenseState = Objects.requireNonNull(licenseState, "licenseState");
+    }
+
+    public void listenForLicenseStateChanges() {
+        /*
+         * Registering this as a listener can not be done in the constructor because otherwise it would be unsafe publication of this. That
+         * is, it would expose this to another thread before the constructor had finished. Therefore, we have a dedicated method to register
+         * the listener that is invoked after the constructor has returned.
+         */
         licenseState.addListener(this);
```
Comment (jpountz) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java]: It'd be nice to have sanity checks that this is called at most once.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java [hunk 2]
```diff
@@ -44,6 +44,14 @@ public OptOutQueryCache(
         this.context = Objects.requireNonNull(context, "threadContext must not be null");
         this.indexName = indexSettings.getIndex().getName();
         this.licenseState = Objects.requireNonNull(licenseState, "licenseState");
+    }
+
+    public void listenForLicenseStateChanges() {
```
Comment (jpountz) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java]: Let's add some minimal javadocs?

**Review by jpountz — APPROVED** [review]
Good catch, how did you find it?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java (hunk 2 above)
Comment (jasontedor) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java]: I pushed 45a2686f3eea4cdd7f72a198fed81ee88b286929.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java (hunk 1 above)
Comment (jasontedor) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/accesscontrol/OptOutQueryCache.java]: I pushed 45a2686f3eea4cdd7f72a198fed81ee88b286929.

**Review by jpountz — APPROVED** [review]

**Review by jaymode — APPROVED** [review]
LGTM. IIRC there are other items like this elsewhere, should these be addressed as a followup?

**Comment by jasontedor** [comment]
> LGTM. IIRC there are other items like this elsewhere, should these be addressed as a followup?

It looks like `LocalExporter` has the same problem with `LicenseStateListener. I will address this in a follow-up indeed.

**Comment by jasontedor** [comment]
>> LGTM. IIRC there are other items like this elsewhere, should these be addressed as a followup?

>It looks like `LocalExporter` has the same problem with `LicenseStateListener`. I will address this in a follow-up indeed.

The `InvalidLicenseEnforcer` has the same issue, so for that I have opened #40985.

**Pull request merged**